File: RB110/small_problems/medium2/lettercase_percentage.rb
=begin
In the easy exercises, we worked on a problem where we had to count the number of uppercase 
and lowercase characters, as well as characters that were neither of those two. Now we want 
to go one step further.

Write a method that takes a string, and then returns a hash that contains 3 entries: 
one represents the percentage of characters in the string that are lowercase letters, 
'one the percentage of characters that are uppercase letters, and one the percentage 
of characters that are neither.

You may assume that the string will always contain at least one character.


Input:
  -String

Output:
  - Hash that represents the percentage of uppercase lower case and neither in the input string

Rules:
    -Explicit
      -Assume that the string will contian at least one character

    -Implicit:
      -the values should be represented by a float
      -White spaces are concidered characters

Questions:

Test Cases // Exsamples:

letter_percentages('abCdef 123') == { lowercase: 50.0, uppercase: 10.0, neither: 40.0 }
letter_percentages('AbCd +Ef') == { lowercase: 37.5, uppercase: 37.5, neither: 25.0 }
letter_percentages('123') == { lowercase: 0.0, uppercase: 0.0, neither: 100.0 }

Data structure:
  -String
  -Hash

assisting Structure
  -Array

Algorithm:
  -Split input string into characters
  -Create an output hash with the different catagories
  -iterate through characters array
    -Check if character is upper case, lowercase or neither
    -Add 1 to each value in the hash depending on the character
  -Transform the values to percentages
  -Return



=end

def find_percentage(num, length)
  num.to_f / length.to_f * 100

end

def letter_percentages(string)
  perc_hash = { :lowercase => 0.0, :uppercase => 0.0, :neither => 0.0}

  perc_hash[:lowercase] = find_percentage(string.count("a-z"), string.length).round(1)
  perc_hash[:uppercase] = find_percentage(string.count("A-Z"), string.length).round(1)
  perc_hash[:neither] = find_percentage(string.delete("a-z").delete("A-Z").length, string.length).round(1)

  perc_hash
end

p letter_percentages('abCdef 123') == { lowercase: 50.0, uppercase: 10.0, neither: 40.0 }
p letter_percentages('AbCd +Ef') == { lowercase: 37.5, uppercase: 37.5, neither: 25.0 }
p letter_percentages('123') == { lowercase: 0.0, uppercase: 0.0, neither: 100.0 }
p letter_percentages('abcdefGHI')

File: LS215/exerises/medium2/letter_case_percentage.js
/*
Write a function that takes a string and returns an object containing the following three properties:

the percentage of characters in the string that are lowercase letters
the percentage of characters that are uppercase letters
the percentage of characters that are neither


Input:
  - A string
Output:
  - An object

Rules
  - The returned object will have the percentages of lower case letters, uppercase letters, neither
  - The strings will include white space and other characters that should be counted as neither
  - We can assume the string will have at least one character?

Questions:
  - Will there always be one argument?
  - Could the length on the input string be any size?
  - Will there be punctuation? If so will it be included in neither?
  - 


Data structure:
  - I may divide the string in to seperate arrays
  - Strings
  - Outputing an object

Algorithm:
  - Create a lowerCase variable and assign to..
    - Match all the elements in the string that are lower case letters
      - Call lengths on the lowerCase return 
  - Create a uppercase variable and assign it..
    - Match all the upperCase letters
      - Call the length on the return
  - Create a neither variable and assign it..
    - Match all the characters in the string that are not letters
     - call length on the return

  - Get the percentages of the three categories
    - percentage = part / total * 100

  return the object { lowercase:  , uppercase: , neither: }
*/

function letterPercentages(string) {
  let lowerCase = string.match(/[a-z]/g) || 0;
  let upperCase = string.match(/[A-Z]/g) || 0;
  let neithers = string.match(/[^a-zA-Z]/g) || 0;

  let lowerCasePerc = ((lowerCase.length / string.length) * 100) || 0;
  let upperCasePerc = ((upperCase.length / string.length) * 100) || 0;
  let neitherPerc = ((neithers.length / string.length) * 100) || 0;

  return { lowercase: `${lowerCasePerc.toFixed(2)}`, uppercase: `${upperCasePerc.toFixed(2)}`, neither: `${neitherPerc.toFixed(2)}`};
}



console.log(letterPercentages('abCdef 123'));
// { lowercase: "50.00", uppercase: "10.00", neither: "40.00" }
console.log(letterPercentages('Hello how are you today?'));

console.log(letterPercentages('AbCd +Ef'));
// { lowercase: "37.50", uppercase: "37.50", neither: "25.00" }

console.log(letterPercentages('123'));
// { lowercase: "0.00", uppercase: "0.00", neither: "100.00" }